### PR TITLE
feat(function): Add fast path for Spark split function

### PR DESCRIPTION
**Pull request by NEUpanning**
Add a fast path that locates the delimiter by iterating string instead of
using re2 library for Spark split function. The microbenchmark shows
there is a 5x performance gain.

```
============================================================================
[...]parksql/benchmarks/SplitBenchmark.cpp     relative  time/iter   iters/s
============================================================================
baseline_old                                                29.14s    34.32m
split_new                                       514.54%      5.66s   176.60m
```

Fixes [#14262](https://github.com/facebookincubator/velox/issues/14262)

### CODE REVIEW

**Comment by NEUpanning** [comment]
@rui-mo Could you help take a look? Thanks!

File: velox/functions/sparksql/Split.h [hunk 1]
```diff
@@ -189,5 +255,8 @@ struct Split {
   }
 
   mutable facebook::velox::functions::detail::ReCache cache_;
+  bool fastPath_{false};
+  char ch_;
```
Comment (rui-mo) [inline velox/functions/sparksql/Split.h]: Would you please add description for the variable `ch_`? It looks not obvious.

File: velox/functions/sparksql/Split.h [hunk 2]
```diff
@@ -72,14 +76,33 @@ struct Split {
...
-      int32_t limit) const {
+      int32_t limit) {
     if (delimiter.empty()) {
       splitEmptyDelimiter(result, input, limit);
     } else {
       split(result, input, delimiter, limit);
     }
   }
 
+  // Octal escaped character in the form \000.
```
Comment (rui-mo) [inline velox/functions/sparksql/Split.h]: Would you please revise the comment? Some example:

> Checks if a string represents an octal-escaped character (e.g., \123). If valid, converts it to its decimal value.

File: velox/functions/sparksql/Split.h [hunk 3]
```diff
@@ -72,14 +76,33 @@ struct Split {
...
+      return false;
+    }
+    for (int i = 1; i < size; ++i) {
+      char ch = str[i];
+      if (ch < '0' || ch > '7') {
+        return false;
+      }
+    }
+    // Octal value must be less than 377.
+    if (std::stoi(str + 1) > 377) {
```
Comment (rui-mo) [inline velox/functions/sparksql/Split.h]: Can the value be equal to 377?

File: velox/functions/sparksql/Split.h [hunk 4]
```diff
@@ -72,14 +76,33 @@ struct Split {
...
+      return false;
+    }
+    decimal = std::stoi(str + 1, nullptr, 8);
```
Comment (rui-mo) [inline velox/functions/sparksql/Split.h:103]: Would the octal value exceed 127, and is it safe to assign it to a char?

File: velox/functions/sparksql/Split.h [hunk 5]
```diff
@@ -113,6 +136,32 @@ struct Split {
     }
   }
 
+  // Fast path if the delimiter is a
+  // (1)one-char String and this character is not one of the RegEx's meta
+  // characters ".$|()[{^?*+\\", or
```
Comment (rui-mo) [inline velox/functions/sparksql/Split.h:142]: Is there any reference for the characters?

File: velox/functions/sparksql/Split.h [hunk 6]
```diff
@@ -113,6 +136,32 @@ struct Split {
...
+    const std::string reservedChars = ".$|()[{^?*+\\";
+    if (delimiterSize == 1) {
+      ch_ = delimiterStr[0];
+      if (reservedChars.find(ch_) == std::string::npos) {
+        fastPath_ = true;
+      }
+    } else if (delimiterSize == 2 && delimiterStr[0] == '\\') {
+      ch_ = delimiterStr[1];
+      if (((ch_ - '0') | ('9' - ch_)) < 0 && ((ch_ - 'a') | ('z' - ch_)) < 0 &&
+          ((ch_ - 'A') | ('Z' - ch_)) < 0) {
```
Comment (rui-mo) [inline velox/functions/sparksql/Split.h]: Would it improve the readability to use `!std::isdigit` and `!std::isalpha` for the checks?

**Review by rui-mo — COMMENTED** [review]
Thanks for this optimization!

File: velox/functions/sparksql/Split.h (hunk 5 above)
Comment (NEUpanning) [reply velox/functions/sparksql/Split.h:142]: In the implementation of JDK, Spark split function utilizes this.
https://github.com/bpupadhyaya/openjdk-8/blob/master/jdk/src/share/classes/java/lang/String.java#L2319

File: velox/functions/sparksql/Split.h (hunk 3 above)
Comment (NEUpanning) [reply velox/functions/sparksql/Split.h:99]: Yes, 377 matches a ÿ character. I will change the comment.

File: velox/functions/sparksql/Split.h (hunk 2 above)
Comment (NEUpanning) [reply velox/functions/sparksql/Split.h]: Thanks!

File: velox/functions/sparksql/Split.h (hunk 4 above)
Comment (NEUpanning) [reply velox/functions/sparksql/Split.h:103]: Thanks! Fast path should only for single character delimiter, so the octal value should not exceed decimal 127. Updated.

Comment (rui-mo) [reply velox/functions/sparksql/Split.h:103]: > if (std::stoi(str + 1) > 177) {

Do you want to check against 127?

File: velox/functions/sparksql/Split.h (hunk 5 above)
Comment (rui-mo) [reply velox/functions/sparksql/Split.h:142]: Would you please add the reference link in the comments?

File: velox/functions/sparksql/Split.h [hunk 7]
```diff
@@ -113,6 +137,33 @@ struct Split {
     }
   }
 
+  // Fast path if the delimiter is a
+  // (1)one-char String and this character is not one of the RegEx's meta
```
Comment (rui-mo) [inline velox/functions/sparksql/Split.h]: nit: use lower case for `String` to be consistent.

File: velox/functions/sparksql/Split.h [hunk 8]
```diff
@@ -113,6 +137,33 @@ struct Split {
...
+  // characters ".$|()[{^?*+\\", or
+  // (2)two-char String and the first char is the backslash and the second is
```
Comment (rui-mo) [inline velox/functions/sparksql/Split.h]: ditto

File: velox/functions/sparksql/benchmarks/SplitBenchmark.cpp [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+BENCHMARK_RELATIVE(split, n) {
+  benchmark->run(n, SplitBenchmark::kSplitExpression);
+}
+
+} // namespace
+
+int main(int argc, char** argv) {
+  folly::Init init{&argc, &argv};
+  memory::MemoryManager::initialize(memory::MemoryManager::Options{});
+  LOG(ERROR) << "Seed: " << seed;
```
Comment (rui-mo) [inline velox/functions/sparksql/benchmarks/SplitBenchmark.cpp]: nit: might be better to use LOG(INFO)

**Review by rui-mo — COMMENTED** [review]
Thanks! Just some nits.

File: velox/functions/sparksql/Split.h (hunk 4 above)
Comment (rui-mo) [reply velox/functions/sparksql/Split.h:103]: It might be helpful to add a test for the decimal value overflow.

Comment (NEUpanning) [reply velox/functions/sparksql/Split.h:103]: > if (std::stoi(str + 1) > 177) {
Do you want to check against 127?

Check against decimal 127, which is octal 177.

**Comment by NEUpanning** [comment]
@rui-mo Thanks for your review. The fuzzer test passed. Here is the command and test statistics.

`spark_expression_fuzzer_test --duration_sec=120 --logtostderr=1 --velox_fuzzer_enable_complex_types --only=split`

```
E20250808 15:02:31.572259 12361364 ExpressionFuzzerVerifier.cpp:259] split 30281 100.00% 918524
E20250808 15:02:31.572293 12361364 ExpressionFuzzerVerifier.cpp:488] Total test cases: 45208
E20250808 15:02:31.572309 12361364 ExpressionFuzzerVerifier.cpp:489] Total test cases verified in the reference DB: 23772
E20250808 15:02:31.574255 12361364 ExpressionFuzzerVerifier.cpp:491] Total test cases failed in both Velox and the reference DB: 21436
E20250808 15:02:31.574281 12361364 ExpressionFuzzerVerifier.cpp:493] Total test cases unsupported in the reference DB: 0
```
